taskboot/cargo.py

```diff
@@ -34,3 +51,31 @@ def cargo_publish(target: Target, args: argparse.Namespace) -> None:
 
     if proc.returncode != 0:
         raise Exception("Failed to publish the crate on crates.io")
+
+
+def cargo_publish(target: Target, args: argparse.Namespace) -> None:
+    """
+    Publish a crate on crates.io
+    """
+
+    # Load config from file/secret
+    config = Configuration(args)
+    assert config.has_cargo_auth(), "Missing Cargo authentication"
+
+    # Check whether the crate to be published must fail or not
+    # in case it has already been published
+    if args.do_not_fail:
+        # Get Cargo.toml content
+        content = open("Cargo.toml").read()
+
+        # Get crate name
+        name = _search_crate_info("name", content, r"^name = \"([^\"]+)\"")
```

Comment:
Instead of trying to find the package and then only uploading it if it's not available, couldn't we attempt to publish, catch the error and ignore it only if it is an "already published" error?

Reply:
Nope, because the error code is the same for many different kinds of errors. We need to find in the error string the substring `already published` to differentiate the error. Our approach is the same one used for `Cargo`, so I think it's pretty good.

Reply:
It won't work if you use a custom registry, I'd prefer using Cargo instead and let it talk to crates.io (or another registry). This way we 1) support publishing to other registries; 2) don't have to deal with changes to the registry API (as we are one level above).
You can either parse the stderr (probably the simplest option), or you can do `cargo search PACKAGENAME` and parse the output to check the latest published version. If it's the same as the current, then it was already published.